File: leetcode-everyday/动态规划/背包问题/416-分割等和子集/416-分割等和子集.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
 var canPartition = function(nums) {
  let sum = nums.reduce((a, b) => a + b);
  if(sum % 2 === 1) return false;
  let target = Math.floor(sum / 2);
  const dp = new Array(target+1).fill(false);
  dp[0] = true;
  for(let i = 1; i <= nums.length; i++) {
      for(let j = target; j >= nums[i-1]; --j) {
          dp[j] = dp[j] || dp[j - nums[i-1]];
      }
  }
  // console.log(dp);
  return dp[target];
};

File: leetcode-everyday/动态规划/背包问题/416-分割等和子集/416-分割等和子集.cpp
class Solution {
public:    
    bool canPartition(vector<int>& nums) {
        if(nums.size() <= 1) return false;
        int sum = accumulate(nums.begin(), nums.end(), 0);
        if(sum % 2 == 1) return false;
        int target = sum / 2, n = nums.size();
        vector<vector<bool>> dp(n+1, vector<bool>(target+1, false));
        dp[0][0] = true;
        for(int i = 1; i <= n; ++i) {
            for(int j = 0; j <= target; ++j) {
                if(j < nums[i-1]) {
                    dp[i][j] = dp[i-1][j];
                } else {
                    dp[i][j] = dp[i-1][j] || dp[i-1][j-nums[i-1]];
                }
            }
        }        
        return dp[n][target];
    }
};


class Solution {
public:
    bool canPartition(vector<int>& nums) {
        if(nums.size() <= 1) return false;
        int sum = accumulate(nums.begin(), nums.end(), 0);
        if(sum % 2 == 1) return false;
        int target = sum / 2, n = nums.size();
        vector<bool> dp(target+1, false);
        dp[0] = true;
        for(int i = 1; i <= n; ++i) {
            for(int j = target; j >= nums[i-1]; --j) {
                dp[j] = dp[j] || dp[j-nums[i-1]];
            }
        }        
        return dp[target];
    }
};
